File: packages/function/src/valuation.ts
import type { Request, Response } from "@google-cloud/functions-framework";
import { Ajv } from "ajv";
import Handlebars from "handlebars";
import { readFileSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { predictValuation } from "./inference.js";

const locations = JSON.parse(
  readFileSync(
    resolve(dirname(fileURLToPath(import.meta.url)), "../locations.json"),
    "utf8",
  ),
) as Record<string, Record<string, Array<string>>>;

const validateSubmission = new Ajv().compile({
  $defs: {
    positiveInteger: {
      anyOf: [{ type: "integer" }, { pattern: "^\\d+$", type: "string" }],
    },
  },
  properties: {
    data: {
      properties: {
        bathrooms: { $ref: "#/$defs/positiveInteger" },
        bedrooms: { $ref: "#/$defs/positiveInteger" },
        city: { maxLength: 100, minLength: 1, type: "string" },
        email: {
          maxLength: 254,
          pattern: "^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$",
          type: "string",
        },
        first_name: { maxLength: 80, type: "string" },
        floor_area: { $ref: "#/$defs/positiveInteger" },
        rates_and_taxes: { $ref: "#/$defs/positiveInteger" },
        property_type: {
          enum: ["Apartment / Flat", "House", "Townhouse"],
          type: "string",
        },
        province: { maxLength: 100, minLength: 1, type: "string" },
        suburb: { maxLength: 100, minLength: 1, type: "string" },
      },
      required: [
        "bathrooms",
        "bedrooms",
        "city",
        "email",
        "floor_area",
        "property_type",
        "province",
        "rates_and_taxes",
        "suburb",
      ],
      type: "object",
    },
    id: { maxLength: 200, minLength: 1, type: "string" },
    status: { const: "completed" },
  },
  required: ["data", "id", "status"],
  type: "object",
});

const templateDirectory = resolve(
  dirname(fileURLToPath(import.meta.url)),
  "../email",
);

const templatesPromise = Promise.all([
  readFile(resolve(templateDirectory, "property-valuation.html"), "utf8").then(
    (template) => Handlebars.compile(template),
  ),
  readFile(resolve(templateDirectory, "property-valuation.txt"), "utf8").then(
    (template) => Handlebars.compile(template, { noEscape: true }),
  ),
]);

const zar = new Intl.NumberFormat("en-ZA", {
  currency: "ZAR",
  maximumFractionDigits: 0,
  style: "currency",
});

export function valuationPresentation(
  estimate: Awaited<ReturnType<typeof predictValuation>>,
): {
  confidenceNote: string;
  headline: string;
  intro: string;
  primaryLabel: string;
  primaryValue: string;
  secondaryLabel: string;
  secondaryValue: string;
} {
  const range = `${zar.format(estimate.low)} – ${zar.format(estimate.high)}`;
  if (estimate.confidence === "low") {
    return {
      confidenceNote: "This is a low-confidence estimate because this property differs from the model’s stronger comparisons. Treat the range cautiously and consider a professional valuation.",
      headline: "Your estimated property value range",
      intro: "The available listing data does not support a reliable single-value estimate for this property.",
      primaryLabel: "Indicative range",
      primaryValue: range,
      secondaryLabel: "",
      secondaryValue: "",
    };
  }
  if (estimate.recommended === null) {
    throw new Error("Point estimate is missing for a non-low confidence tier");
  }
  if (estimate.confidence === "high") {
    return {
      confidenceNote: "This estimate falls within the model’s higher-confidence group, but it remains an automated asking-price estimate rather than a formal valuation.",
      headline: "Your estimated property value",
      intro: "Based on the details you shared, your property’s recommended value is:",
      primaryLabel: "Recommended value",
      primaryValue: zar.format(estimate.recommended),
      secondaryLabel: "Likely range",
      secondaryValue: range,
    };
  }
  return {
    confidenceNote: "This estimate has moderate confidence, so the likely range is more reliable than the central estimate alone.",
    headline: "Your estimated property value range",
    intro: "Based on the details you shared, the model estimates this likely range:",
    primaryLabel: "Likely range",
    primaryValue: range,
    secondaryLabel: "Central estimate",
    secondaryValue: zar.format(estimate.recommended),
  };
}

export async function valuation(
  request: Request,
  response: Response,
): Promise<void> {
  response.set("Cache-Control", "no-store");

  if (request.method !== "POST") {
    response.set("Allow", "POST");
    response.status(405).json({ error: "Method not allowed" });
    return;
  }

  const submission = { ...request.body, data: { ...request.body?.data } };
  const invalidFirstName =
    typeof submission.data.first_name === "string" &&
    submission.data.first_name !== "" &&
    submission.data.first_name.trim() === "";

  for (const [field, value] of Object.entries(submission.data)) {
    if (typeof value === "string") {
      submission.data[field] = value.trim();
    }
  }

  submission.id =
    typeof submission.id === "string" ? submission.id.trim() : submission.id;

  if (invalidFirstName || !validateSubmission(submission)) {
    response.status(400).json({ error: "Submission is invalid" });
    return;
  }

  const property = {
    bathrooms: Number(submission.data.bathrooms),
    bedrooms: Number(submission.data.bedrooms),
    locality_1: submission.data.city,
    locality_2: submission.data.suburb,
    rates_and_taxes: Number(submission.data.rates_and_taxes),
    region: submission.data.province,
    size: Number(submission.data.floor_area),
    type: submission.data.property_type,
  };

  if (
    [
      [property.bathrooms, 1, 20],
      [property.bedrooms, 1, 20],
      [property.rates_and_taxes, 1, 100000],
      [property.size, 10, 5000],
    ].some(
      ([value, minimum, maximum]) =>
        !Number.isInteger(value) || value < minimum || value > maximum,
    ) ||
    !locations[property.region]?.[property.locality_1]?.includes(
      property.locality_2,
    )
  ) {
    response.status(400).json({ error: "Submission is invalid" });
    return;
  }

  let estimate;

  try {
    estimate = await predictValuation(property);
  } catch {
    response.status(500).json({ error: "Valuation failed" });
    return;
  }

  try {
    const apiKey = process.env.RESEND_API_KEY;
    const from = process.env.RESEND_FROM_EMAIL;

    if (!apiKey || !from) {
      throw new Error("Resend is not configured");
    }

    const [toHtml, toText] = await templatesPromise;
    const presentation = valuationPresentation(estimate);
    const values = {
      ...presentation,
      bathrooms: property.bathrooms,
      bedrooms: property.bedrooms,
      greeting: submission.data.first_name
        ? `Hi ${submission.data.first_name},`
        : "Hello,",
      location: [
        property.locality_2,
        property.locality_1,
        property.region,
      ].join(", "),
      size: `${property.size.toLocaleString("en-ZA")} m²`,
      type: property.type,
    };
    const result = await fetch("https://api.resend.com/emails", {
      body: JSON.stringify({
        from,
        html: toHtml(values),
        subject: "Your South African property value estimate",
        text: toText(values),
        to: [submission.data.email],
      }),
      headers: {
        Authorization: `Bearer ${apiKey}`,
        "Content-Type": "application/json",
        "Idempotency-Key": `property-valuation/${submission.id}`,
      },
      method: "POST",
    });

    if (!result.ok) {
      throw new Error(`Resend rejected the email with status ${result.status}`);
    }
  } catch {
    response.status(502).json({ error: "Email delivery failed" });
    return;
  }

  response.status(204).send();
}

File: packages/model/valuation.py
"""Form submission → validated inputs → rendered email → Resend delivery."""
import json
import math
import os
import re
from pathlib import Path
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from jinja2 import Environment, FileSystemLoader, StrictUndefined, select_autoescape

DIRECTORY = Path(__file__).resolve().parent
LOCATIONS = json.loads((DIRECTORY / "locations.json").read_text())
TEMPLATES = Environment(
    loader=FileSystemLoader(DIRECTORY / "email"),
    autoescape=select_autoescape(["html"]),
    undefined=StrictUndefined,
    keep_trailing_newline=True,
)


def parse_submission(body: object) -> dict:
    """Keep form validation separate from the model's more permissive input contract."""
    if not isinstance(body, dict) or body.get("status") != "completed" or not isinstance(body.get("data"), dict):
        raise ValueError("Submission is invalid")
    identifier = body.get("id")
    if not isinstance(identifier, str) or not 1 <= len(identifier.strip()) <= 200 or re.search(r"[\r\n]", identifier):
        raise ValueError("Submission ID is invalid")
    data = {key: value.strip() if isinstance(value, str) else value for key, value in body["data"].items()}
    first_name = body["data"].get("first_name", "")
    if not isinstance(first_name, str) or (first_name and not first_name.strip()) or len(first_name.strip()) > 80:
        raise ValueError("First name is invalid")
    email = data.get("email")
    if not isinstance(email, str) or len(email) > 254 or not re.fullmatch(r"[^\s@]+@[^\s@]+\.[^\s@]+", email):
        raise ValueError("Email is invalid")
    property_ = {}
    for field in ("province", "city", "suburb"):
        value = data.get(field)
        if not isinstance(value, str) or not 1 <= len(value) <= 100:
            raise ValueError("Location is invalid")
        property_[field] = value
    if property_["suburb"] not in LOCATIONS.get(property_["province"], {}).get(property_["city"], []):
        raise ValueError("Location is invalid")
    for source, target, minimum, maximum in (
        ("bedrooms", "bedrooms", 1, 20),
        ("bathrooms", "bathrooms", 1, 20),
        ("floor_area", "floor_size", 10, 5000),
        ("rates_and_taxes", "rates", 1, 100000),
    ):
        value = data.get(source)
        if isinstance(value, str) and re.fullmatch(r"[0-9]+", value):
            value = int(value)
        if (isinstance(value, bool) or not isinstance(value, (int, float))
                or not minimum <= value <= maximum or value != int(value)):
            raise ValueError(f"{source} is invalid")
        property_[target] = int(value)
    return dict(id=identifier.strip(), email=email, first_name=first_name.strip(), property=property_)


def render_email(submission: dict, estimate: dict) -> dict:
    """Display the model's rounded recommendation directly; escape HTML, not plain text."""
    property_ = submission["property"]
    price = estimate["recommended_asking_price_zar"]
    if not math.isfinite(price) or price < 0 or price % 1000:
        raise ValueError("Model returned an invalid rounded asking price")
    values = dict(
        greeting=f"Hi {submission['first_name']}," if submission["first_name"] else "Hello,",
        headline="Your recommended asking price",
        intro="Based on the details you shared, your property's recommended asking price is:",
        primaryLabel="Recommended asking price",
        primaryValue="R\u00a0" + f"{price:,.0f}".replace(",", "\u00a0"),
        estimateNote="This is an automated asking-price estimate based on advertised listings, not a formal valuation or a prediction of the final sale price.",
        location=", ".join(property_[key] for key in ("suburb", "city", "province")),
        bedrooms=property_["bedrooms"],
        bathrooms=property_["bathrooms"],
        size=f"{property_['floor_size']:,}".replace(",", "\u00a0") + " m²",
    )
    return dict(
        to=[submission["email"]],
        subject="Your South African property value estimate",
        html=TEMPLATES.get_template("property-valuation.html").render(values),
        text=TEMPLATES.get_template("property-valuation.txt").render(values),
    )


def send_email(identifier: str, email: dict) -> None:
    """Resend deduplicates retries using the original submission ID."""
    api_key = os.environ.get("RESEND_API_KEY")
    sender = os.environ.get("RESEND_FROM_EMAIL")
    if not api_key or not sender:
        raise RuntimeError("Resend is not configured")
    request = Request(
        "https://api.resend.com/emails",
        data=json.dumps({**email, "from": sender}).encode(),
        headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json",
                 "Idempotency-Key": f"property-valuation/{identifier}"},
        method="POST",
    )
    try:
        with urlopen(request, timeout=30) as response:
            if not 200 <= response.status < 300:
                raise RuntimeError(f"Resend rejected the email with status {response.status}")
    except HTTPError as error:
        error.close()
        raise RuntimeError(f"Resend rejected the email with status {error.code}") from error
